### ISSUE TITLE
关于模型

### ISSUE THREAD

**Issue by WenTingTseng** [xhw205/GlobalPointer_torch#2]
你好:
看来一下您的原始码
似乎是以BERT作为Encoder然后在BERT模型上再加Global Pointer的部分。是否有模型架构图可以参考，方便做理解呢?

谢谢

**Comment by xhw205** [comment]
gp的本质是 token-pair 做实体首尾打分，具体的就是：实体首部向量si，尾部向量sj，然后是si与sj做类似self-attention乘法得到【batchsize*cls_num*length*length】的logits分数。si、sj就是由bert做encoder得到的，这个encoder可以随意更换，换成lstm，cnn啥的都可以。

**Comment by WenTingTseng** [comment]
那再请教一个问题，它是如何知道哪边是实体首部，哪边是实体尾部呢

**Comment by xhw205** [comment]
把【cls_num×length×length】的logits分数想象成一个三维矩阵，【length×length】代表矩阵的长宽，【cls_num】矩阵的厚度。假设【length×length】矩阵中的(2,3)下标的得分大于阈值，2就代表实体的头部token下标，3就代表实体的尾部token下标。就反推回去句子抽取实体就行了

**Comment by WenTingTseng** [comment]
OK。那实体首部向量si，尾部向量sj，然后是si与sj做类似self-attention乘法得到【batchsize×cls_num×length×length】的logits分数。这个产生首位部向量具体代码应该如下 首部向量就是qw  尾部向量是kw吧。一开始进入到torch.split的outputs是BERT当作encoder产生的向量。我想请教代码中这三行，最终切分产生 qw和kw过程的逻辑? 为什么qw是首部向量，kw是尾部向量
`outputs = torch.split(outputs, self.inner_dim * 2, dim=-1)`
` outputs = torch.stack(outputs, dim=-2)`
 `qw, kw = outputs[...,:self.inner_dim], outputs[...,self.inner_dim:]`

另外，"假设【length×length】矩阵中的(2,3)下标的得分大于阈值" 这个阈值作法是在代码如下吗?
`# 排除下三角`
   `mask = torch.tril(torch.ones_like(logits), -1)#torch.tril()函数可以获取张量下三角部分的内容，而将上三角部分的元素设置为0`
   `logits = logits - mask * 1e12`
   `return logits/self.inner_dim**0.5`

感谢回覆

**Comment by xhw205** [comment]
[这是GP作者博客](https://kexue.fm/archives/8373)
一、我的第一次解释的“实体首部向量si，尾部向量sj做类self-attention“，指的是token-pair范式下的通用具体操作，参考BiaffineNER、multi-head等等NER模型，并不是GP的具体操作...可能随手回复，没做好断句...
二、GP不用两个单独的模块去分别识别实体的头和尾，**qw, kw认为是头尾向量是不正确的哈**（很抱歉初始回复的误导），它俩就是个变换矩阵，有点双仿射的意思，目的就为了生成【B×NUM×L×L】的逻辑得分，这种乘性attention表示token-pair得分挺常见的，也没有个真正的好解释
三、阈值是推理/eval时候才需要设定的，默认就是0，见predict_CME.py中`for l, start, end in zip(*np.where(scores > 0))`
四、排除下三角的目的就是，防止实体尾巴在头部前面，就是说只保证 “正向实体”，见[参考图](https://kexue.fm/usr/uploads/2021/05/2377306125.png)

**Comment by WenTingTseng** [comment]
所以qw, kw是两个变换矩阵，这边双仿射的操作是不是如下图所示
![雙仿射](https://user-images.githubusercontent.com/32416416/165870920-bc87e22c-d4f7-4971-8322-4c85e6a22c59.JPG)

我们这边的Encoder是BERT模型，同样会通过一层MLP后产生两个向量qw, kw。然后qw, kw在算内积产生分数。或是说我们这个操作也有点像下图呢
![v2-7c2ea2a221998339cbc995e08043dd7a_b](https://user-images.githubusercontent.com/32416416/165870463-5ad83462-fd5f-45e8-9517-95d1f2e5de5d.jpg)
那GP的操作应该是在哪里?这部分我不太理解，是计算那些分数吗?然后透过分数去指向该选哪一个(Start,end)吗

另外，在您的程式码内multilabel_categorical_crossentropy函式参数传入第一个是y_pred，第二个是y_true
`def multilabel_categorical_crossentropy(y_pred, y_true):`
但是真实在使用参数传入第一个是y_true，第二个是y_pred
`multilabel_categorical_crossentropy(y_true, y_pred)`

这部分我有点困惑，须向您再请教

謝謝

**Comment by xhw205** [comment]
这俩图是用两个模块分别识别实体的首和尾。GlobalPointer 将首尾视为一个整体去进行判别，具体操作就是 直接通过torch.split 后分出的两个向量计算内积作为打分。建议您仔细读一下原文，模型细节可以在原文下与[原作者交流](https://kexue.fm/archives/8373)
第二部分，没啥问题，变量名写混了，不影响啥

**Comment by WenTingTseng** [comment]
好的，谢谢你
